gson/src/main/java/com/google/gson/Gson.java

```diff
@@ -1360,7 +1358,19 @@ public <T> T fromJson(JsonReader reader, TypeToken<T> typeOfT)
       JsonToken unused = reader.peek();
       isEmpty = false;
       TypeAdapter<T> typeAdapter = getAdapter(typeOfT);
-      return typeAdapter.read(reader);
+      T object = typeAdapter.read(reader);
+      Class<?> expectedTypeWrapped = Primitives.wrap(typeOfT.getRawType());
```

Comment:
It seems all other `fromJson` methods delegate to this one, so I have added the type check (which was previously `Primitives.wrap(classOfT).cast(object)`) here instead.
However, this method did not actually have this check before.

Reply:
I ran this change against all of Google's internal tests and found a _lot_ of failures with this exception, or with an exception about not being able to instantiate the abstract class `com.google.common.collect.ImmutableList`. In both cases I think the code in question should be fixed, because it is relying on accidental behaviour. A pattern I saw quite often was this:

```java
@AutoValue
public abstract Foo {
  public abstract ImmutableList<Bar> bars();

  public static TypeAdapter<Foo> typeAdapter(Gson gson) {
    return new AutoValue_Foo.GsonTypeAdapter(gson);
  }

  @AutoValue.Builder
  public interface Builder {
    public Builder setBars(List<Bar> bars);
    public Foo build();
  }
}
```

This is using [auto-value-gson](https://github.com/rharter/auto-value-gson) to serialize and deserialize AutoValue classes. It works even if no `TypeAdapterFactory` has been registered for `ImmutableList`, because Gson deserializes a plain `List`. Auto-value-gson constructs the `Foo` instance using its builder, and the deserialized `List` works with the `setBars` method. That method calls `ImmutableList.copyOf(List)` so we do get the right `ImmutableList` type. But the additional checks in this PR don't let us get that far.

As I say, code like that only works accidentally. My plan is to add a `GuavaCollectionsTypeAdapterFactory` to `extras` and update the internal code to use that. I already found several custom `ImmutableListTypeAdapterFactory` and `ImmutableMapTypeAdapterFactory` classes which could be switched to use this.

Some of the other failures were just plain using the wrong types, which again happened to work through erasure and the like.

So in short, I think this change is OK, but it may require some client code to be updated. Most of the updates should be straightforward and will lead to client code that is more correct.

I think it will be a while before we can complete this because of all the adjustments that will need to be made.